ocaml/typing/ctype.ml

```diff
@@ -1581,17 +1581,29 @@ let prim_mode mvar = function
 let rec instance_prim_locals locals mvar macc finalret ty =
   match locals, get_desc ty with
   | l :: locals, Tarrow ((lbl,marg,mret),arg,ret,commu) ->
-     let marg = Mode.Alloc.with_locality (prim_mode (Some mvar) l) marg in
+      let marg' = Alloc.newvar () in
+      Locality.equate_exn (Alloc.locality marg') (prim_mode (Some mvar) l);
+      Alloc.submode_exn marg' (Alloc.set_locality_max marg);
+      Alloc.submode_exn (Alloc.set_locality_min marg) marg';
+      let marg = marg' in
      let macc =
-       Mode.Alloc.join [mret;
-         Mode.Alloc.close_over marg;
-         Mode.Alloc.partial_apply macc
+       Alloc.join [Alloc.disallow_right mret;
+         Alloc.close_over marg.comonadic marg.monadic;
```

Comment:
`close_over` and `partial_apply` have slightly different requirement on left/right status - we would lose some generality if we unify the two.

Reply:
But do we need that generality? It looks to me that they're always used together, so that suggests that we don't need the extra generality, even if it is possible.

It's also possible that with crisp comments describing what these do, they'll make sense separately. Right now, I don't have enough of a grasp on them to be sure.

Reply:
With the new comments, the way forward becomes clear. See https://github.com/riaqn/flambda-backend/pull/4

Reply:
Thanks - I suggest that we merge the current PR first, and discuss the proposed changes. As I think the discussion could take some time.

I created internal ticket 2300 for future reference.

Reply:
That's fine.